src/suites/cts/validation.spec.ts

```diff
@@ -0,0 +1,673 @@
+export const description = `
+Validation tests.
+`;
+
+import { TestGroup } from '../../framework/index.js';
+
+import { ValidationTest } from './validation_test.js';
+
+export const g = new TestGroup(ValidationTest);
+
+g.test('validation/binding count mismatch', async t => {
+  const bindGroupLayout = t.device.createBindGroupLayout({
+    bindings: [
+      {
+        binding: 0,
+        visibility: GPUShaderStage.COMPUTE,
+        type: 'storage-buffer',
+      },
+    ],
+  });
+
+  t.device.createBindGroup({
+    bindings: [],
+    layout: bindGroupLayout,
+  });
+
+  await t.expectUncapturedError();
+});
+
+g.test('validation/binding must be present in layout', async t => {
+  const bindGroupLayout = t.device.createBindGroupLayout({
+    bindings: [
+      {
+        binding: 0,
+        visibility: GPUShaderStage.COMPUTE,
+        type: 'storage-buffer',
+      },
+    ],
+  });
+
+  t.device.createBindGroup({
+    bindings: [
+      {
+        binding: 1,
+        resource: {
+          buffer: t.storageBuffer,
+        },
+      },
+    ],
+    layout: bindGroupLayout,
+  });
+
+  await t.expectUncapturedError();
+});
+
+g.test('validation/binding index too high in bind group', async t => {
+  const bindingIndexTooHigh = 16;
+
+  const bindGroupLayout = t.device.createBindGroupLayout({
+    bindings: [
+      {
+        binding: 0,
+        visibility: GPUShaderStage.COMPUTE,
+        type: 'storage-buffer',
+      },
+    ],
+  });
+
+  t.device.createBindGroup({
+    bindings: [
+      {
+        binding: bindingIndexTooHigh,
+        resource: {
+          buffer: t.storageBuffer,
+        },
+      },
+    ],
+    layout: bindGroupLayout,
+  });
+
+  await t.expectUncapturedError();
+});
+
+g.test('validation/same binding cannot be set twice', async t => {
+  const bindGroupLayout = t.device.createBindGroupLayout({
+    bindings: [
+      {
+        binding: 0,
+        visibility: GPUShaderStage.COMPUTE,
+        type: 'storage-buffer',
+      },
+      {
+        binding: 1,
+        visibility: GPUShaderStage.COMPUTE,
+        type: 'storage-buffer',
+      },
+    ],
+  });
+
+  t.device.createBindGroup({
+    bindings: [
+      {
+        binding: 0,
+        resource: {
+          buffer: t.storageBuffer,
+        },
+      },
+      {
+        binding: 0,
+        resource: {
+          buffer: t.storageBuffer,
+        },
+      },
+    ],
+    layout: bindGroupLayout,
+  });
+
+  await t.expectUncapturedError();
+});
+
+g.test('validation/storage buffer binding must contain exactly one storage buffer', async t => {
```

Comment:
Can you add a TODO for me above this: `TODO(kainino0x): Look into whether it's possible to parameterize this set of tests.`